### ISSUE TITLE
Mapping between HLSL and GLSL matrix conventions

### ISSUE THREAD

**Issue by BearishSun** [LukasBanana/XShaderCompiler#39]
HLSL matrices are named as "row X column", i.e. `float3x4` has 3 rows and 4 columns. 
GLSL matrices on the other hand are named as "column X row", i.e. `mat3x4` has 3 columns and 4 rows. 

This same naming scheme also applies to swizzles and array indices when accessing matrices, as well as the order in which the initializers interpret their arguments.

Currently the translation between matrix types is by directly mapping their indices. This code doesn't compile when translated to GLSL:

```
float3x4 mat;
vec4 vec;
vec3 result = mul(mat, vec);
```

Because the matrix is translated to `mat3x4`, and multiplication of a 3-column and 4-row matrix with a 4-column vector is not defined.

-------

I can see two possible solutions to the problem above:
 1. Keep the matrices as they are, but require external code to transpose them before passing them to GLSL. Also switch the order of arguments of any `mul` intrinsics.
 2. Map the matrices to their row/column equivalents in GLSL, flip swizzle access indices, flip array access indices and re-map initializers.

I have actually already started working on option 2 but am running into a problem that I'm not sure how to solve. 

In particular the problem arises when we need to translate assignment to a single matrix row `mat[1] = float4(1, 2, 3, 4);` into GLSL, where this needs to map to multiple accesses (one for each column), e.g. 
```
mat4x3 mat;

// Write to second row
vec4 tmp = vec4(1, 2, 3, 4);
mat[0][1] = tmp[0];
mat[1][1] = tmp[1];
mat[2][1] = tmp[2];
mat[3][1] = tmp[2];
```

I can do this fine for `ExprStmnt`, but what is the right way to handle this when such an assignment occurs in a loop conditional? Such matrix assignments are unlikely to be used in real world, so for now I just plan to print out errors, but if possible I'd like to solve it properly, eventually.

Because if this problem I am considering that maybe we should go with option 1 instead (transposing matrices & reversing multiply order). What do you think?

**Comment by BearishSun** [comment]
Now that I wrote it down I had an idea. 

Auto-generate functions that handle the row writes. Detect when they're used and then generate a function like `mat3x4 writeToMatRow(inout mat3x4 mat, int row, float4 val)`. It should then be possible to just replace any such assignment expression with a function call instead. Sounds valid?

**Comment by LukasBanana** [comment]
Option 1 is definitely the easier option, and I would consider option 2 only later when the compiler has other issues solved first.

I'm aware of the issue you mentioned about a loop-condition/loop-iterator, and I solved this for geometry shaders with the `EmitVertex` function by assigning everything inside a comma separated `SequenceExpr` but it looks reeeeally ugly:
```glsl
for (; ; gl_Position = output[0].position, xsv_COLOR = output[0].color, EmitVertex()) {}
```
And yes, this is a case which might never happen in an actual program, but a sophisticated compiler should also handle such rare cases.

Your idea with the function call is the same I had with the `clip` intrinsic.
A wrapper function for `clip` is only generated if it appears in a sequence:
```glsl
// First "clip" intrinsic
if (alpha < 0.5)
    discard;

// Second and third "clip" intrinsics
xsc_clip(alpha - 0.5), xsc_clip(alpha - 0.5);
```

Now before you start chaging the argument order of all `mul` intrinsics, I'm not sure if this is ok for quadratic matrices, because `mul(wvpMatrix, v)` in HLSL has always resulted in the same as `wvpMatrix * v` in GLSL so far, at least in my 3D frameworks.

**Comment by BearishSun** [comment]
I'm glad to see you had similar problems (and already solved them).

I'm probably going to play with option 2 for a bit, and commit it on a local branch if it's ever needed (I think I have a good idea on how to do it now).  

Regarding 3D frameworks, you'll need to change them so they transpose the matrices. Otherwise the behavior will not be the same. This goes for both square and non-square matrices. I also need to change Banshee, which is a bit of a pain because I still mix generated GLSL with manually written GLSL, and manually written code all assumes column vectors (multiply by matrix from the left).

**Comment by LukasBanana** [comment]
As far as I know: GLSL matrices are considered to be **column-major matrices** (which is common in mathematics) and also uses **column-major storage**.
HLSL matrices are considered to be **row-major matrices** but also uses **row-major storage**.
This should compensate for the differences, i.e. I always pass my matrices as column-majors matrix/storage but it works for both GLSL and HLSL.

```
4x4 Matrix in column-major (GLSL):
/ x0 y0 z0 w0 \
| x1 y1 z1 w1 |
| x2 y2 z2 w2 |
\ x3 y3 z3 w3 /

4x4 Matrix in flattened column-major storage (GLSL):
ArrayGLSL = [ x0, x1, x2, x3, y0, y1, y2, y3, z0, z1, z2, z3, w0, w1, w2, w3 ]

4x4 Matrix in row-major (HLSL):
/ x0 x1 x2 x3 \
| y0 y1 y2 y3 |
| z0 z1 z2 z3 |
\ w0 w1 w2 w3 /

4x4 Matrix in flattened row-major storage (HLSL):
ArrayHLSL = [ x0, x1, x2, x3, y0, y1, y2, y3, z0, z1, z2, z3, w0, w1, w2, w3 ]
```
As you can see, `ArrayGLSL == ArrayHLSL`. That's why it always worked for 4x4 matrices at least.

**Comment by BearishSun** [comment]
This is all getting confusing :)

In your example the HLSL matrix is a transpose of the GLSL matrix. Which means that in HLSL it would be `mul(mat, vec)` and equivalent in GLSL `vec * mat`. So even if we change nothing on the 3D framework side, flipping the `mul` argument should work. Fine so far.

What confuses me is that you said you use the same multiplication order in both HLSL and GLSL, but using the logic above that shouldn't work!?

Also note that GLSL matrices inside a block with a `std140` qualifier (which seems to be the default for XShaderCompiler) use row-major storage. Taken from GLSL spec.:
```
layout(std140) uniform Transform { // layout of this block is std140
 mat4 M1; // row_major
 layout(column_major) mat4 M2; // column major
 mat3 N1; // row_major
};
```
  
--------

Instead of getting bogged down into all these confusing layers of matrix ordering and storage I'm instead just going by the simple rule: flipping the mul arguments means all matrices need to be transposed. It doesn't matter which is their original ordering, storage or size.

**Comment by LukasBanana** [comment]
The HLSL matrix in my example is transposed, because this is the default layout for an HLSL matrix: row-major.

Anyway, you mean you want to change the matrix order and require the users to transpose the input matrices, but isn't that equivalent to changing nothing and don't require the users to transpose anything (at least for 4x4 matrices)?
Because changing the argument order of the `mul` Intrinsic in HLSL and the matrix multiplication operator in GLSL is equivalent to transposing one matrix from the argument list.

Before I start with such a breaking change, I will check how `glslang` and `SPIRV-Cross` make this translation.

And yes, this is indeed a confusing issue between Direct3D and OpenGL.

**Comment by BearishSun** [comment]
Yep, for 4x4 matrices it is the equivalent of doing nothing as far as multiplication goes. But it fixes the problems with the non-square matrices which **must** to have their mul arguments flipped (and therefore need to be transposed).

But transposing a matrix fixes the swizzle and array access problem as well, not only multiplication. Imagine we leave square matrices as is:
 - Their swizzles and array indices are still wrong. i.e. `mat._m23` in HLSL will map to row 2, col 3, while in GLSL it will map to col 2, row 3. Since we didn't transpose the matrix, this will yield different values in GLSL than it does in HLSL.

Also it makes it unclear how to multiply a (transposed) 3x4 matrix with a (non-transposed) 4x4 matrix.

Looking it up in other solutions sounds like a good idea.

**Comment by LukasBanana** [comment]
Ok, I think I now understand you.
I never had problems with it because I never use non-square matrices.
But you're right, it's not a solution to keep it as it is for non-square matrices.

So yeah, I will have a look at `glslang` and `SPRIV-Cross`.
Also `DirectXShaderCompiler` looks promising as it is about to get a code generator for SPIR-V IR.

**Comment by LukasBanana** [comment]
You were right: `glslang` and `SPIRV-Cross` change the argument order of the `mul` intrinsic in the GLSL output, but only for matrices, not for vectors.
For vectors, `mul(a, b)` becomes `dot(a, b)`, which is also not the case for the XShaderCompiler!
Moreover, `glslang` always changes the layout qualifiers from `row_major` to `column_major` and vice versa.
The swizzle operators seem to be equal.

Therefore, there is some more work to do for the translation of `mul` intrinsics in the XShaderComiler.
I'll take care of this, but you can of course also do some research if you like.

Thanks for pointing this out.

**Comment by BearishSun** [comment]
That sounds good to me! Changing the layout qualifiers achieves the same effect as transposing the matrix externally (I think), so that's a nice solution to the problem.

I did actually go ahead and implement option 2 (https://github.com/BearishSun/XShaderCompiler/tree/matrix-flip). It transforms this:
```
cbuffer Test
{
	float3x4 m;
	float3x4 m2[3];
};

float4 main() : SV_Position
{
	m[2][3] = 0.0f;
	m._m23 = 5.0f;
	m2[1][2][3] = 1.0f;

	float4 row = m[1];
	float val = m[2].r;
	m2[2][1] = row;
	
	const float3x3 cmat =
		{
			0.8951f,  0.2664f, -0.1614f,
			-0.7502f,  1.7135f,  0.0367f,
			0.0389f, -0.0685f,  1.0296f,
		};	
	
	const float3x3 cmat2 =
		{
			float3(0.8951f,  0.2664f, -0.1614f),
			float3(-0.7502f,  1.7135f,  0.0367f),
			float3(0.0389f, -0.0685f,  1.0296f)
		};		
	
    return m._m23 + row;
}
```
into this:

```
vec4 xsc_matWriteRow(inout mat4x3 m, int i, vec4 v)
{
    m[0][i] = v[0];
    m[1][i] = v[1];
    m[2][i] = v[2];
    m[3][i] = v[3];
    return v;
}

vec4 xsc_matReadRow(mat4x3 m, int i)
{
    return vec4(m[0][i], m[1][i], m[2][i], m[3][i]);
}

mat3 xsc_matConstruct(vec3 v0, vec3 v1, vec3 v2)
{
    return mat3(v0[0], v1[0], v2[0], v0[1], v1[1], v2[1], v0[2], v1[2], v2[2]);
}

layout(std140) uniform Test
{
    mat4x3 m;
    mat4x3 m2[3];
};

void main()
{
    m[3][2] = 0.0f;
    m._m32 = 5.0f;
    m2[1][3][2] = 1.0f;
    vec4 row = vec4(xsc_matReadRow(m, 1));
    float val = xsc_matReadRow(m, 2).r;
    xsc_matWriteRow(m2[2], 1, row);
    mat3 cmat = mat3(0.8951f, -0.7502f, 0.0389f, 0.2664f, 1.7135f, -0.0685f, -0.1614f, 0.0367f, 1.0296f);
    mat3 cmat2 = xsc_matConstruct(vec3(0.8951f, 0.2664f, -0.1614f), vec3(-0.7502f, 1.7135f, 0.0367f), vec3(0.0389f, -0.0685f, 1.0296f));
    gl_Position = vec4(m._m32) + row;
}

```

But seeing how `glslang` and `SPRIV-Cross` go with the other approach it gives me more confidence in option 1. Initially I felt it was hacky, but it does yield cleaner code and potentially better performance because the row accesses can be mapped directly.

I'll continue using my version in order to get me through the compile errors I'm having now, but will replace it with your version when its done.

**Comment by LukasBanana** [comment]
Alright, that's a good idea.
Maybe I can start with this next week.

**Comment by LukasBanana** [comment]
This should be fixed with [79d5863](https://github.com/LukasBanana/XShaderCompiler/commit/79d58639e2fb0760deb290a430745480428b6346):

* The order of arguments for _all_ `mul` intrinsic calls are now _always_ swapped.
* The type modifiers `row_major` and `column_major` are swapped for all matrix types that appear inside a `UniformBufferDecl` (i.e. only supported for GLSL `version >= 140`), for lower GLSL versions, the client programmer is required to change the storage layout on its own when passing them to OpenGL.
* The `mul` intrinsic class with vector-vector multiplication is converted to a `dot` intrinsic call.

Just reopen this issue when you still have problems with it.

**Comment by BearishSun** [comment]
Nice, thank you.
